### ISSUE TITLE
[XGrid] Document how to use the scroll feature

### ISSUE THREAD

**Issue by oliviertassinari** [mui-org/material-ui-x#1103]
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

This feature is already implemented but the documentation doesn't cover it. From the eye of a developer, it's as if we had never implemented it. I propose we add a section for it. 

## Examples 🌈

- https://adazzle.github.io/react-data-grid/canary/?path=/story/demos--scroll-to-row
- https://www.ag-grid.com/documentation/react/grid-api/#reference-scrolling

I'm not sure we need an actual demo for the feature. I think that the main problem is that there is no way to use this feature without diving deep into the source. 

## Motivation 🔦

One use case example: "as a developer, I'm streaming updates to the rows in the data grid. When an update is pushed, I would like to scroll the data grid to make it visible."

**Comment by embeddedt** [comment]
I'd also like to know if there is a way of doing this for DataGrid (the MIT version). The documentation mentioned that there are supposed to be prop equivalents for things in `apiRef`, but I'm having trouble finding this for `scrollToIndexes`. I looked for something relating to scrolling state or a scrolling model but couldn't find anything.

**Comment by oliviertassinari** [comment]
@embeddedt I agree with the general feeling. We use virtualization in the DataGrid but curb it. We prevent developers to use the regular DOM API to scroll to a specific index, It seems that this API should be available to developers, the same way it's available in https://react-window.vercel.app/#/examples/list/scroll-to-item as an example.

**Comment by cactusjosh** [comment]
I've been able to interact with X-Grid's `useGridApiRef()` hook to do something like this:
```javascript
  const tableApiRef = useGridApiRef();

  // inside a useEffect()
 const columnIndexToScrollTo = tableApiRef.current.getRowIndex(someIDField);
 tableApiRef.current.scrollToIndexes(columnIndexToScrollTo);
``` 
But I can't seem to get it to work. Since there's no documentation (I'm even guessing the fn parameters) I have no idea what I'm doing wrong.

**Comment by oliviertassinari** [comment]
Related to https://github.com/mui-org/material-ui-x/issues/1818

**Comment by embeddedt** [comment]
@cactusjosh Thanks for the example! Unfortunately, this doesn't help me since I'm using DataGrid and not XGrid, which means that I don't have access to the apiRef functions.

**Comment by oliviertassinari** [comment]
> Thanks for the example! Unfortunately, this doesn't help me since I'm using DataGrid and not XGrid, which means that I don't have access to the apiRef functions.

@embeddedt Ok, this should be solved with #1781

**Comment by embeddedt** [comment]
@oliviertassinari Maybe I'm missing something, but that feels very unidiomatic for React. It seems that I would have to do something like this (semi-pseudocode):

```jsx
const gridRef = useRef(null);
const scrollToRow = useCallback((i) => {
    const rowEl = gridRef.current.querySelector(`div[data-id="${i}"]`);
    if(rowEl != null)
        rowEl.scrollIntoView();
}, [ gridRef ]);
return <DataGrid {...otherProps} ref={gridRef}/>;
```

This seems like it could easily break if the underlying approach for rendering rows change. Ideally I would like to just call an API that hides this implementation detail and scrolls to the desired index.

Please let me know if I misinterpreted the description.

**Comment by oliviertassinari** [comment]
@embeddedt Yes, this is the approach that we could encourage, with an extra `role="row"`. I would expect it to be good enough.

**Comment by ZeeshanTamboli** [comment]
@oliviertassinari I can work on this. How do we approach? Points below:
1. Should we have a new section altogether? eg: `DataGrid - Scrolling` path: `components/data-grid/scrolling`
2. Add it in a existing section/page.

You are saying:
> I'm not sure we need an actual demo for the feature

Should we not have a demo like https://virtuoso.dev/scroll-to-index/? What other possible ways you mean to document without a demo?

**Comment by oliviertassinari** [comment]
@ZeeshanTamboli I believe this feature is linked to virtualization. Without virtualization, developers could use the web [.scrollIntoView](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) API without caring about what the data grid has available. So https://material-ui.com/components/data-grid/virtualization/ with a new h2.

Regarding how we would document it. I would propose we have at the minimum an api Ref section like https://material-ui.com/components/data-grid/filtering/#apiref. And if we want to go the extra mile, a live demo.

> AFAIK it is accessible in apiRef only, right?

Correct

**Comment by ZeeshanTamboli** [comment]
I just found out that we have the scroll APIs documented here at GridApi: https://material-ui.com/api/data-grid/grid-api/

@oliviertassinari 
Still would it be better to have it documented it again separately on virtualizaton section as part of virtualizaton  like filtering API ref  https://material-ui.com/components/data-grid/filtering/#apiref ?

**Comment by oliviertassinari** [comment]
@ZeeshanTamboli Correct Yes.

**Comment by nkennedy21** [comment]
scrollToIndexes only works properly if the grid height is exactly (n rows * rowHeight) + headerHeight. this is because of the transform: translate3d sytyle attribute in the renderingZone. making the grid height dynamic is more difficult because of this (still possible with calculations making it the height of a specific number of rows plus the headerHeight)

**Comment by m4theushw** [comment]
Closed by #2634

**Comment by luan-dang-techlabs** [comment]
> * https://www.ag-grid.com/documentation/react/grid-api/#reference-scrolling



> I've been able to interact with X-Grid's `useGridApiRef()` hook to do something like this:
> 
> ```js
> const tableApiRef = useGridApiRef();
> 
> // inside a useEffect()
> const columnIndexToScrollTo = tableApiRef.current.getRowIndex(someIDField);
> tableApiRef.current.scrollToIndexes(columnIndexToScrollTo);
> ```
> 
> But I can't seem to get it to work. Since there's no documentation (I'm even guessing the fn parameters) I have no idea what I'm doing wrong.
> 
> Edit 1. I was using `scrollToIndexes` incorrectly. Here's what I think you need instead:
> 
> ```js
> const columnIndexes = {rowIndex: 1, colIndex: 1};
> tableApiRef.current.scrollToIndexes(columnIndexToScrollTo);
> ```

I wrapped the call in  requestAnimationFrame() to make scrollToIndexes() work.

```
 useEffect(() => {
    const animate = () => {
         // rowIndex: 20 can be anything
         const columnIndexes = {rowIndex: 20, colIndex: 1};
        tableApiRef.current.scrollToIndexes(columnIndexToScrollTo);
    }

// allow next repaint to make scrollToIndexes method to scroll to 'rowIndex'
    const id = requestAnimationFrame(animate);

     return () => {
   cancelAnimationFrame(id);
}
 }, [someDependecies, tableApiRef]);
```

**Comment by luan-dang-techlabs** [comment]
> scrollToIndexes only works properly if the grid height is exactly (n rows * rowHeight) + headerHeight. this is because of the transform: translate3d sytyle attribute in the renderingZone. making the grid height dynamic is more difficult because of this (still possible with calculations making it the height of a specific number of rows plus the headerHeight)

@nkennedy21 - Can you elaborate on this? I think what you say make sense, I had to wrap the scrollToIndexes() call in a requestAnimationFrame() to make it work.

**Comment by c3-anerishah** [comment]
@m4theushw it looks like you'd added the documentation for scrolling to https://mui.com/x/react-data-grid/scrolling/ based on your comment above, but the demo doesn't seem to be working - clicking on Bottom doesn't scroll to the bottom of the grid and same with the other button respectively. I was wondering if something has changed in the `scrollToIndexes` API?
